src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,177 @@
+}
+
+SidechainTxsCommitmentBuilder::EmptyField SidechainTxsCommitmentBuilder::emptyField;
+
+#ifdef BITCOIN_TX
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx) { return; }
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert) { return; }
+uint256 SidechainTxsCommitmentBuilder::getCommitment() { return uint256(); }
+#else
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx)
+{
+    if (!tx.IsScVersion())
+        return;
+
+    unsigned int ccOutIdx = 0;
+    LogPrint("sc", "%s():%d -getting leaves for vsc out\n", __func__, __LINE__);
+    for(; ccOutIdx < tx.GetVscCcOut().size(); ++ccOutIdx)
+    {
+        uint256 scId = tx.GetVscCcOut().at(ccOutIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVscCcOut()[ccOutIdx].GetHash(), tx.GetHash(), ccOutIdx);
+        vec.push_back(pField);
+    }
+
+    LogPrint("sc", "%s():%d -getting leaves for vft out\n", __func__, __LINE__);
+    for(; ccOutIdx < tx.GetVftCcOut().size(); ++ccOutIdx)
+    {
+        uint256 scId = tx.GetVftCcOut().at(ccOutIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVftCcOut()[ccOutIdx].GetHash(), tx.GetHash(), ccOutIdx);
+        vec.push_back(pField);
+    }
+
+    LogPrint("sc", "%s():%d - nIdx[%d]\n", __func__, __LINE__, ccOutIdx);
+}
+
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert)
+{
+    uint256 scId = cert.GetScId();
+    sScIds.insert(scId);
+    mScCerts[scId] = mapCertToField(cert.GetHash());
+}
+
+uint256 SidechainTxsCommitmentBuilder::getCommitment()
+{
+    std::vector<field_t*> vSortedScLeaves;
+
+    // set of scid is ordered
+    for (const auto& scid : sScIds)
+    {
+        field_t* ftRootField = nullptr;
+        auto itFt = mScMerkleTreeLeavesFt.find(scid);
+        if (itFt != mScMerkleTreeLeavesFt.end() )
+        {
+            ftRootField = merkleTreeRootOf(itFt->second);
+        }
+
+        field_t* btrRootField = nullptr;
+        auto itBtr = mScMerkleTreeLeavesBtr.find(scid);
+        if (itBtr != mScMerkleTreeLeavesBtr.end() )
+        {
+            btrRootField = merkleTreeRootOf(itBtr->second);
+        }
+
+        field_t* certRootField = nullptr;
+        auto itCert = mScCerts.find(scid);
+        if (itCert != mScCerts.end() )
+        {
+            certRootField = itCert->second;
+        }
+
+        field_t* scIdField = mapCertToField(scid);
+
+        std::vector<field_t*> singleSidechainComponents = {ftRootField, btrRootField, certRootField, scIdField};
+        field_t * sidechainTreeRoot = merkleTreeRootOf(singleSidechainComponents);
+
+        vSortedScLeaves.push_back(sidechainTreeRoot);
+    }
+
+    field_t * finalTreeRoot = emptyField.zero;
```

Comment:
Probably it's best to actually compute the root even if all leaves are zero (or set it to a phantom root, or the actual hash of all 0 nodes up until the root). Keep in mind always the "SNARK friendliness" we want to achieve